proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1352,6 +1354,544 @@ lemma scheduleUsed_corres:
+              valid_objs'
+              (handle_overrun_loop usage)
+              (handleOverrunLoop usage')"
+  apply (rule_tac Q="\<lambda>s'. sc_at' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: sc_at_cross simp: state_relation_def)
+  apply (rule_tac Q="\<lambda>s'. is_active_sc' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: is_active_sc'_cross simp: state_relation_def)
+  apply (clarsimp simp: handle_overrun_loop_def handleOverrunLoop_def runReaderT_def)
+  apply (rule corres_whileLoop_abs; simp?)
+       apply (frule_tac usage=r' in head_time_buffer_equiv; simp?)
+       apply fastforce
+      apply (corressimp corres: handleOverrunLoopBody_corres)
+     apply (wpsimp wp: handle_overrun_loop_body_no_fail)
+     apply (clarsimp simp: vs_all_heap_simps)
+    apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+        apply (rule hoare_weaken_pre)
+         apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+          apply wpsimp
+         apply wpsimp
+        apply wpsimp
+       apply (rule hoare_weaken_pre)
+        apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+         apply wpsimp
+        apply wpsimp
+       apply wpsimp
+      apply (rule hoare_weaken_pre)
+       apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+        apply wpsimp
+       apply wpsimp
+      apply wpsimp
+     apply (rule hoare_weaken_pre)
+      apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+       apply (wpsimp wp: handle_overrun_loop_body_non_zero_refills)
+      apply wpsimp
+     apply clarsimp
+    apply (rule hoare_weaken_pre)
+     apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+      apply (wpsimp wp: handle_overrun_loop_body_refills_unat_sum_equals_budget)
+     apply wpsimp
+    apply clarsimp
+   apply (clarsimp simp: pred_conj_def)
+   apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+     apply (wpsimp wp: handleOverrunLoopBody_valid_objs')
+    apply (rule hoare_weaken_pre)
+     apply (rule_tac f=ksCurSc in hoare_lift_Pf)
+      apply wpsimp
+     apply wpsimp
+    apply wpsimp
+   apply (rule hoare_weaken_pre)
+    apply (rule_tac f=ksCurSc in hoare_lift_Pf)
+     apply wpsimp
+    apply wpsimp
+   apply wpsimp
+  apply (fastforce intro:  handle_overrun_loop_body_terminates)
+  done
+
+lemma handle_overrun_loop_is_active_sc[wp]:
+  "handle_overrun_loop usage \<lbrace>\<lambda>s. is_active_sc sc_ptr s\<rbrace>"
+  apply handle_overrun_loop_simple
+  done
+
+lemma get_refills_exs_valid[wp]:
+  "sc_at sc_ptr s \<Longrightarrow> \<lbrace>(=) s\<rbrace> get_refills sc_ptr \<exists>\<lbrace>\<lambda>r. (=) s\<rbrace>"
+  apply (clarsimp simp: get_refills_def)
+  apply (wpsimp wp: get_sched_context_exs_valid)
+   apply (erule sc_atD1)
+  apply simp
+  done
+
+(* FIXME RT: keep? move to Lib? *)
+lemma corres_when3:
+  "\<lbrakk>G = G'; G \<Longrightarrow> corres_underlying sr nf nf' dc P P' a c\<rbrakk>
+   \<Longrightarrow> corres_underlying sr nf nf' dc P P' (when G a) (when G' c)"
```

Comment:
I don’t think we’d want to keep adding these weaker lemmas, to be honest. Surely you can replace this with `corres_if_split` and a few more steps?

Reply:
Removed